### ISSUE TITLE
Syntax for ending a section

### ISSUE THREAD

**Issue by infinity0** [sphinx-doc/sphinx#8518]
I want to do something like this:

```rst
==========
Chapter A
==========

Intro
========

Introduction text

.. toctree::
   :hidden:
   :glob:
   :maxdepth: 1

   a/*
```

Sphinx today will structure the `a/*` pages underneath the `Intro` section, but instead I want them directly underneath the top-level `Chapter A` section, i.e. on the same level as the `Intro` subsection.

This seems not to be possible in sphinx today (e.g. the answers on https://stackoverflow.com/questions/60877841/how-can-i-force-sphinx-to-end-a-chapter-when-making-a-pdf are basically "you can't"), and is very frustrating.

**Comment by infinity0** [comment]
One suggested alternative, moving the toctree above the intro and marking it `hidden`, also moves the subsection headings *above* the intro structurally, and this is apparent when rendering the TOC, the intro ends up *below* everything else. This is obviously also not acceptable.

**Comment by tk0miya** [comment]
Unfortunately, the syntax of reStructuredText is not ours. So we can't change it. Additionally, I guess what you want is a globbing-include feature. So I recommend you to develop your own custom include directive supporting `:glob:` option.

**Comment by infinity0** [comment]
@tk0miya are you saying that it is impossible to implement this feature in a directive? Is that because reST parsing of sections comes before directives? In that case I have filed a ticket to docutils here: https://sourceforge.net/p/docutils/bugs/411/

**Comment by tk0miya** [comment]
If my understanding is correct (because I'm not an expert of the reST parser!), the reST parser invokes a directive to create nodes as the part of the content of the current section. Therefore, directives can't control the parent section of the generated nodes.

**Comment by infinity0** [comment]
The rST developers are unwilling to add syntax for explicitly ending a section, because the validity rules in rST forbid body elements after a subsection is closed (in the parent section).

In the case of the toctree directive however, it only adds sections so this is fine in terms of document validity. The rST maintainers suggest:

> Instead, the `<section>`-generating directive could define a directive option
that can be used to specify whether the generated `<section>` node should be nested in the current section (becoming a sub-section) or appended to the current section (becoming the next section of the same level).

I.e. it seems that rST provides the ability for sphinx to choose whether to nest or append the generated node, which sphinx can make use of for this feature.

**Comment by tk0miya** [comment]
According to the design of the toctree directive, I feel odd to add such an option. The purpose of the directive is to build a part of "Table of Content" tree by connecting other pages as sub-pages of the current location. So I object to your suggestion.

**Comment by infinity0** [comment]
"The current location" is what is at debate here - there is currently no way to say "the current location" should be (1) "inside the current section" or (2) "after the current section", sphinx assumes (1) but the user could want (2).

So I don't entirely understand your objection.

What alternative do you suggest, how about a `siblingtoctree` directive in that case? But the code to implement it will be nearly the same as `toctree`.

**Comment by jakobandersen** [comment]
I see the point of the request, but I think it would be difficult to implement, and it is not clear to me how it would work in general:

- The toctree directive returns a lists of nodes for the doctree, so I guess an implementation would need to do some potentially complicate post-transform for the doctree to achieve this.
- What should the result be if there is content after the toctree? It would be strange if the previous section (here "Intro") just continues.

If the "Intro" section is short, I suggest removing the section heading. If it is long, I suggest splicing it out to another file and put it in the toctree. Then all the sections will be at the same level.

**Comment by infinity0** [comment]
> The toctree directive returns a lists of nodes for the doctree, so I guess an implementation would need to do some potentially complicate post-transform for the doctree to achieve this.

According to the rST developer I spoke to (see the bug report linked before), this is just the difference between nesting vs appending the directive contents to the current node list - presumably rST gives directives the ability to do this, otherwise the developer wouldn't have mentioned this as an option?

> What should the result be if there is content after the toctree? It would be strange if the previous section (here "Intro") just continues.

I think continuing the previous section is just fine. If I understand correctly, the toctree directive as rendered in the document, is just a normal body display element that is part of the previous section already (in the current version of the code, without implementing this feature), and any content after that is also part of the previous section already, with or without enabling this feature.

In other words, this feature would only affect the overall website structure as interpreted by sphinx.

**Comment by jakobandersen** [comment]
> > The toctree directive returns a lists of nodes for the doctree, so I guess an implementation would need to do some potentially complicate post-transform for the doctree to achieve this.
> 
> According to the rST developer I spoke to (see the bug report linked before), this is just the difference between nesting vs appending the directive contents to the current node list - presumably rST gives directives the ability to do this, otherwise the developer wouldn't have mentioned this as an option?

Right, but the toctree directive does not know it's parent node yet (it may not even exist), hence this would have to be a post-transform.

> 
> > What should the result be if there is content after the toctree? It would be strange if the previous section (here "Intro") just continues.
> 
> I think continuing the previous section is just fine. If I understand correctly, the toctree directive as rendered in the document, is just a normal body display element that is part of the previous section already (in the current version of the code, without implementing this feature), and any content after that is also part of the previous section already, with or without enabling this feature.
> 
> In other words, this feature would only affect the overall website structure as interpreted by sphinx.

There could also be additional sections and subsections so the post-transform, either the toctree would insert after the current section or after all sections in the level of the current section, but on the surface both of these seems a bit icky.

Note that html is just one output format. It needs to make sense for all output formats, and potentially also when the toctree is not hidden. I'm not sure how, e.g., the Latex output looks when hidden toctrees are used.

**Comment by tk0miya** [comment]
>What alternative do you suggest, how about a siblingtoctree directive in that case? But the code to implement it will be nearly the same as toctree.

I think adding a new directive is a better way. I'll consider merging your directive to sphinx-core if other users will favor it.
